crates/edr_instrument/src/coverage.rs

```diff
@@ -0,0 +1,212 @@
+    pub source: String,
+    pub metadata: Vec<InstrumentationMetadata>,
+}
+
+/// Computes a deterministic hash for a statement in a Solidity file.
+///
+/// For the time being, we're assuming that compilation configuration has no
+/// impact on the statement hash. This is a simplification, but is considered a
+/// reasonable trade-off for the current use case.
+fn compute_deterministic_hash_for_statement(
+    source_id: &str,
+    content_hash: &B256,
+    statement_counter: u64,
+) -> B256 {
+    let mut hasher = Keccak256::new();
+
+    hasher.update(source_id);
+    hasher.update(content_hash);
+    hasher.update(statement_counter.to_le_bytes());
+
+    let hash = hasher.finalize();
+    B256::new(hash.into())
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum InstrumentationError {
+    #[error(transparent)]
+    Initialization(#[from] ParserInitializationError),
+    #[error("Failed to parse Solidity file.")]
+    ParseError { errors: Vec<ParseError> },
+}
+
+pub fn instrument_code(
+    source_code: &str,
+    source_id: &str,
+    solidity_version: Version,
+) -> Result<InstrumentationResult, InstrumentationError> {
+    let parser = SolidityParser::create(solidity_version)?;
+    let parse_result = parser.parse(SolidityParser::ROOT_KIND, source_code);
+    if !parse_result.is_valid() {
+        return Err(InstrumentationError::ParseError {
+            errors: parse_result.errors().clone(),
+        });
+    }
+
+    // Compute the content hash once
+    let content_hash = B256::new(Keccak256::digest(source_code).into());
+
+    let mut instrumented_source = String::new();
+    let mut metadata = Vec::new();
+
+    let mut statement_counter: u64 = 0;
+    let mut cursor = parse_result.create_tree_cursor();
+    while cursor.go_to_next() {
+        match cursor.node() {
+            Node::Nonterminal(node) if node.kind == NonterminalKind::Statement => {
+                statement_counter += 1;
+
+                let hash = compute_deterministic_hash_for_statement(
+                    source_id,
+                    &content_hash,
+                    statement_counter,
+                );
+
+                instrumented_source.push_str(&format!("__HardhatCoverage.sendHit({hash}); "));
+
+                let range = cursor.text_range();
+                metadata.push(InstrumentationMetadata {
+                    tag: hash,
+                    kind: "statement",
+                    start_utf16: range.start.utf16,
+                    end_utf16: range.end.utf16,
+                });
+            }
+            Node::Terminal(node) => {
+                instrumented_source.push_str(&node.text);
+            }
+            Node::Nonterminal(_) => {}
+        }
+    }
+
+    instrumented_source.push_str("\nimport \"hardhat/coverage.sol\";");
+
+    Ok(InstrumentationResult {
+        source: instrumented_source,
+        metadata,
+    })
+
+    // serde_json::to_string_pretty(&metadata)
+    //.context("Failed to serialize metadata")?
```

Comment:
Leftover?

Reply:
Good catch. Addressed in https://github.com/NomicFoundation/edr/pull/879/commits/86e6e12ae68fc8fcedd45b0b6f7e7e41dba5eca7